src/x509.lisp

```diff
@@ -0,0 +1,225 @@
+(in-package :cl+ssl)
+
+#|
+ASN1 string validation references:
+ - https://github.com/digitalbazaar/forge/blob/909e312878838f46ba6d70e90264650b05eb8bde/js/asn1.js
+ - http://www.obj-sys.com/asn1tutorial/node128.html
+ - https://github.com/deadtrickster/ssl_verify_hostname.erl/blob/master/src/ssl_verify_hostname.erl
+ - https://golang.org/src/encoding/asn1/asn1.go?m=text
+|#
+(defgeneric decode-asn1-string (asn1-string type))
+
+(defun copy-bytes-to-lisp-vector (src-ptr vector count)
+  (declare (type (simple-array (unsigned-byte 8)) vector)
+           (type fixnum count)
+           (optimize (safety 0) (debug 0) (speed 3)))
+  (dotimes (i count vector)
+    (setf (aref vector i) (cffi:mem-aref src-ptr :unsigned-char i))))
+
+(defun asn1-string-bytes-vector (asn1-string)
+  (let* ((data (asn1-string-data asn1-string))
+         (length (asn1-string-length asn1-string))
+         (vector (cffi:make-shareable-byte-vector length)))
+    (copy-bytes-to-lisp-vector data vector length)
+    vector))
+
+(declaim (inline asn1-iastring-char-p))
+(defun asn1-iastring-char-p (byte)
+  (declare (type (unsigned-byte 8) byte)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (< byte #x80))
+
+(defun asn1-iastring-p (bytes)
+  (declare (type (simple-array (unsigned-byte 8)) bytes)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (every #'asn1-iastring-char-p bytes))
+
+(defmethod decode-asn1-string (asn1-string (type (eql +v-asn1-iastring+)))
+  (let ((bytes (asn1-string-bytes-vector asn1-string)))
+    (if (asn1-iastring-p bytes)
+        (babel:octets-to-string bytes :encoding :ascii)
```

Comment:
This would require to declare :babel as an explicit dependency in our ASDF system. But as we already depend on flexi-streams, we could use `flexi-streams:octets-to-string`